src/sentry/models/deploy.py

```diff
@@ -23,8 +25,70 @@ class Deploy(Model):
     date_started = models.DateTimeField(null=True, blank=True)
     name = models.CharField(max_length=64, null=True, blank=True)
     url = models.URLField(null=True, blank=True)
-    notified = models.NullBooleanField(null=True, db_index=True)
+    notified = models.NullBooleanField(null=True, db_index=True, default=False)
 
     class Meta:
         app_label = 'sentry'
         db_table = 'sentry_deploy'
+
+    @staticmethod
+    def get_lock_key(deploy_id):
+        return 'deploy-notify:%s' % deploy_id
+
+    @classmethod
+    def notify_if_ready(cls, deploy_id, fetch_complete=False):
+        """
+        create activity and send deploy notifications
+        if they haven't been sent
+        """
+        from sentry.models import Activity, Environment, ReleaseCommit, ReleaseHeadCommit
+
+        lock_key = cls.get_lock_key(deploy_id)
+        lock = locks.get(lock_key, duration=5)
```

Comment:
We're doing a lot of stuff in this lock, we _might_ need more than 5 seconds in some edge cases. I'd bump this to 30 just to be safe to avoid overrunning the lock and sending two emails.